### PR TITLE
[WiP] CreateInteraction/PerformInteraction: case with the same MenuName or vrCommands between the different ChoiceSets

### PR DESCRIPTION
**Pull request by LevchenkoS**
CreateInteractionChoiceSet: SUCCESS the same menuName or vrCommand in different choiceSets

---
Functional requirement: [APPLINK-16940](https://adc.luxoft.com/jira/browse/APPLINK-16940)
Implements: [APPLINK-29256](https://adc.luxoft.com/jira/browse/APPLINK-29256)
Related to: [APPLINK-29259](https://adc.luxoft.com/jira/browse/APPLINK-29259)

---
Unit tests has been updated
for `CreateInteractionRequest`.

---
Related to: [APPLINK-29262](https://adc.luxoft.com/jira/browse/APPLINK-29262)

PerformInteraction: DUPLICATE_NAME MenuName or/and vrSynonyms within ChoiceSets
-
Functional requirement: [APPLINK-17385](https://adc.luxoft.com/jira/browse/APPLINK-17385)
Implements: [APPLINK-29256](https://adc.luxoft.com/jira/browse/APPLINK-29256)
Related to: [APPLINK-29259](https://adc.luxoft.com/jira/browse/APPLINK-29259)

---
Unit tests has been updated
for `PerformInteractionRequest`.

---
Related to: [APPLINK-29262](https://adc.luxoft.com/jira/browse/APPLINK-29262)
Recreated #923 

@AGaliuzov, @Kozoriz, @LuxoftAKutsan, @VProdanov, @wolfylambova22, @VVeremjova, @dtrunov, please, review.

### CODE REVIEW

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 1]
```diff
@@ -267,12 +344,33 @@ void CreateInteractionChoiceSetRequest::SendVRAddCommandRequests(
...
 
-    msg_params[strings::cmd_id] =
-        choice_set[strings::choice_set][chs_num][strings::choice_id];
-    msg_params[strings::vr_commands] =
-        smart_objects::SmartObject(smart_objects::SmartType_Array);
-    msg_params[strings::vr_commands] =
-        choice_set[strings::choice_set][chs_num][strings::vr_commands];
+    const SmartObject& choice = choice_set[strings::choice_set][chs_num];
+
+    if (!choice[strings::vr_commands].asArray() ||
```
Comment (AGaliuzov) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @LevchenkoS 
`!choice[strings::vr_commands].asArray() ` looks quite strange.
Ok `asArray` returns pointer but is is not clear from this call.
I would suggest you to split ths logic

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h [hunk 2]
```diff
@@ -234,6 +238,38 @@ class PerformInteractionRequest : public CommandRequestImpl {
   DISALLOW_COPY_AND_ASSIGN(PerformInteractionRequest);
 };
 
+template <typename FunctorT>
+bool PerformInteractionRequest::ProcessChoiceSet(
```
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @AGaliuzov, just a visual one. Nothing more than it

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 3]
```diff
@@ -184,77 +233,99 @@ bool CreateInteractionChoiceSetRequest::compareStr(
...
+  for (; vr_end != vr_it; ++vr_it) {
+    const std::string vr_command = (*vr_it).asString();
+    result = utils::InsertIntoSet(utils::Djb2HashFromString(vr_command),
+                                  vr_commands_hash) &&
+             result;
+  }
+  return result;
+}
+
+std::set<int32_t> CollectVRCommands(
```
Comment (Kozoriz) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @LevchenkoS , please add typedef for set.

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 4]
```diff
@@ -267,12 +344,33 @@ void CreateInteractionChoiceSetRequest::SendVRAddCommandRequests(
...
-        choice_set[strings::choice_set][chs_num][strings::choice_id];
-    msg_params[strings::vr_commands] =
-        smart_objects::SmartObject(smart_objects::SmartType_Array);
-    msg_params[strings::vr_commands] =
-        choice_set[strings::choice_set][chs_num][strings::vr_commands];
+    const SmartObject& choice = choice_set[strings::choice_set][chs_num];
+
+    if (!choice[strings::vr_commands].asArray() ||
+        choice[strings::vr_commands].empty()) {
+      --expected_chs_count_;
```
Comment (wolfylambova22) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @LevchenkoS I see you not the one named this variable like this, but if is not a problem for you could you rename it to be more meaningful ?

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc [hunk 5]
```diff
@@ -114,11 +180,10 @@ TEST_F(CreateInteractionChoiceSetRequestTest, OnTimeout_GENERIC_ERROR) {
   utils::SharedPtr<CreateInteractionChoiceSetRequest> req_vr =
       CreateCommand<CreateInteractionChoiceSetRequest>();
 
-  MockAppPtr mock_app = CreateMockApp();
-  EXPECT_CALL(app_mngr_, application(_)).WillOnce(Return(mock_app));
-  ON_CALL(*mock_app, app_id()).WillByDefault(Return(kConnectionKey));
-  ON_CALL(*mock_app, get_grammar_id()).WillByDefault(Return(kConnectionKey));
-  ON_CALL(*mock_app, RemoveCommand(_)).WillByDefault(Return());
+  EXPECT_CALL(app_mngr_, application(_)).WillOnce(Return(mock_app_));
```
Comment (wolfylambova22) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @LevchenkoS why do you skip this argument?

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc [hunk 6]
```diff
@@ -205,7 +256,539 @@ TEST_F(CreateInteractionChoiceSetRequestTest, OnEvent_VR_UNSUPPORTED_RESOURCE) {
...
+  Event event(hmi_apis::FunctionID::VR_AddCommand);
+
+  (*message_)[am::strings::params][am::strings::correlation_id] =
+      kCorrelationId;
+  (*message_)[am::strings::params][am::hmi_response::code] =
+      hmi_apis::Common_Result::eType::INVALID_DATA;
+
+  FillMessageFieldsItem1(message_);
+  FillMessageFieldsItem2(message_);
+  EXPECT_CALL(app_mngr_, application(_)).WillOnce(Return(mock_app_));
```
Comment (wolfylambova22) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @LevchenkoS kConnectionKey again? If I am correct, please check all expectations..

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc [hunk 7]
```diff
@@ -205,7 +256,539 @@ TEST_F(CreateInteractionChoiceSetRequestTest, OnEvent_VR_UNSUPPORTED_RESOURCE) {
...
+  EXPECT_CALL(*mock_app_, AddChoiceSet(kChoiceSetId, _));
+  ON_CALL(app_mngr_, GetNextHMICorrelationID())
+      .WillByDefault(Return(kCorrelationId));
+  command_->Run();
+  EXPECT_CALL(app_mngr_, updateRequestTimeout(_, _, _));
+  EXPECT_CALL(app_mngr_, TerminateRequest(_, _, _)).Times(0);
+  event.set_smart_object(*message_);
+
+  command_->on_event(event);
+}
```
Comment (wolfylambova22) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @LevchenkoS don't we expect call for ManageMobileCommand?

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 8]
```diff
@@ -108,42 +142,63 @@ void CreateInteractionChoiceSetRequest::Run() {
...
-    std::pair<std::set<uint32_t>::iterator, bool> ins_res =
-        choice_id_set.insert((*choice_set_it)[strings::choice_id].asInt());
-    if (!ins_res.second) {
+    const uint32_t choice_id = (*choice_set_it)[strings::choice_id].asUInt();
+    if (!utils::InsertIntoSet(choice_id, choice_id_set)) {
       LOG4CXX_ERROR(logger_,
-                    "Choise with ID "
-                        << (*choice_set_it)[strings::choice_id].asInt()
-                        << " already exists");
+                    "Choise with ID " << choice_id << " already exists");
```
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @wolfylambova22, actually it should be `Choice ID`. It's ID of a `Choice`, not a chosen ID

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h [hunk 9]
```diff
@@ -234,6 +238,38 @@ class PerformInteractionRequest : public CommandRequestImpl {
...
+    const application_manager::ApplicationSharedPtr app, FunctorT checker) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using smart_objects::SmartObject;
+  using smart_objects::SmartArray;
+  SmartObject& choice_set_list =
```
Comment (dtrunov) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @LevchenkoS I think better name is choice_set_id_list.

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc [hunk 10]
```diff
@@ -205,7 +256,539 @@ TEST_F(CreateInteractionChoiceSetRequestTest, OnEvent_VR_UNSUPPORTED_RESOURCE) {
...
+
+  smart_objects::SmartObject* choice_set_id = NULL;
+  EXPECT_CALL(*mock_app_, FindChoiceSet(kChoiceSetId))
+      .WillRepeatedly(Return(choice_set_id));
+  EXPECT_CALL(app_mngr_, application(_)).WillRepeatedly(Return(mock_app_));
+
+  EXPECT_CALL(*mock_message_helper_, VerifyImage(_, _, _))
+      .WillRepeatedly(Return(mobile_apis::Result::SUCCESS));
+
+  if ((*message_)[am::strings::msg_params][am::strings::choice_set][0]
```
Comment (VProdanov) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @LevchenkoS is this check necessary , it looks redundant.

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @VProdanov, Fixed in 1f27dd6

Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @VProdanov, fixed in f677eaf

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc (hunk 6 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @wolfylambova22, fixed in f677eaf

File: src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc (hunk 5 above)
Comment (LevchenkoS) [inline src/components/application_manager/test/commands/mobile/create_interaction_choice_set_test.cc]: @wolfylambova22, fixed in f677eaf

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc (hunk 4 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @wolfylambova22, I rather leave it as it is.

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc (hunk 3 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @Kozoriz, fixed in 	85f94b7

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc (hunk 1 above)
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @AGaliuzov , fixed in 85f94b7

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 11]
```diff
@@ -184,77 +239,99 @@ bool CreateInteractionChoiceSetRequest::compareStr(
...
+
+  SmartArray::const_iterator vr_it = vr_commands_array->begin();
+  const SmartArray::const_iterator vr_end = vr_commands_array->end();
+
+  bool result = true;
+  for (; vr_end != vr_it; ++vr_it) {
+    const std::string vr_command = (*vr_it).asString();
+    result = utils::InsertIntoSet(utils::Djb2HashFromString(vr_command),
+                                  vr_commands_hash) &&
+             result;
```
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @AGaliuzov , fixed in 85f94b7

File: src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc [hunk 12]
```diff
@@ -184,77 +239,99 @@ bool CreateInteractionChoiceSetRequest::compareStr(
...
+          logger_,
+          "Invalid syntax of secondary_image value. Syntax check failed");
       return true;
     }
   }
   return false;
 }
 
+bool InsertVrCommands(const smart_objects::SmartObject& vr_commands,
+                      std::set<int32_t>& vr_commands_hash) {
```
Comment (LevchenkoS) [inline src/components/application_manager/src/commands/mobile/create_interaction_choice_set_request.cc]: @AGaliuzov , fixed in 85f94b7

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h [hunk 13]
```diff
@@ -87,6 +89,16 @@ class PerformInteractionRequest : public CommandRequestImpl {
    */
   virtual void onTimeOut();
 
+#if BUILD_TESTS
```
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @AGaliuzov, fixed in 6dfae28

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h [hunk 14]
```diff
@@ -131,26 +143,18 @@ class PerformInteractionRequest : public CommandRequestImpl {
...
+   * @param checker a functor which checks choises;
    *
-   * return Return TRUE if there are no similar VR synonyms in choice set,
+   * return Return TRUE if checker returned TRUE
+   * for each choice in a choice set form the list,
    * otherwise FALSE
    */
-  bool CheckChoiceSetVRSynonyms(
-      application_manager::ApplicationSharedPtr const app);
+  template <typename FunctorT>
```
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @AGaliuzov, fixed in 6dfae28

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h (hunk 2 above)
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @AGaliuzov, fixed in 6dfae28

File: src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h (hunk 9 above)
Comment (LevchenkoS) [inline src/components/application_manager/include/application_manager/commands/mobile/perform_interaction_request.h]: @dtrunov , fixed in 6dfae28

**Comment by LuxoftAKutsan** [comment]
@LevchenkoS what this pull request is waiting for?

**Comment by LuxoftAKutsan** [comment]
@LevchenkoS  please check base. Probably it should not be develop

**Comment by LuxoftAKutsan** [comment]
@LevchenkoS are you sur that this changes required in develop?

**Comment by LevchenkoS** [comment]
PR is not actual

**Pull request closed without merging**